In [74]:
import torch
import numpy as np
import matplotlib.pyplot as plt
import torchvision.transforms as transforms


from torch import nn
from torchvision.datasets import MNIST
from torch.utils.data import random_split, DataLoader
import torch.optim as optim

In [75]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

print("Using device:", device)

Using device: cpu


# Cargar y particionar el MNIST dataset

In [76]:
mnist_train = MNIST(root = 'data/', train = True, transform = transforms.ToTensor(), download = True)
mnist_test = MNIST(root = 'data/', train = False, transform = transforms.ToTensor())

# Creación de clase de red neuronal + función de entrenamiento

In [77]:
# Clase de Red Neuronal #1
class FeedForwardNN(nn.Module):
  def __init__(self, input_size, hidden_size, output_size):
    super().__init__()
    self.hidden = nn.Linear(input_size, hidden_size)
    self.output = nn.Linear(hidden_size, output_size)
    self.relu = nn.ReLU()

  def forward(self, x):
    x = x.reshape(-1, 784)
    x = self.hidden(x)
    x = self.relu(x)
    x = self.output(x)
    return x

In [78]:
def calc_accuracy(y_pred, y_true):
  nan, y_prediction = torch.max(y_pred, 1)
  correct_predictions = torch.sum(y_prediction == y_true).item()
  accuracy = correct_predictions / len(y_pred)
  return accuracy

In [79]:
def train(model, num_epochs, optimizer, batch_size):
  train_loader = DataLoader(mnist_train, batch_size = batch_size, shuffle = True)
  val_loader = DataLoader(mnist_train, batch_size = 10000, shuffle = True)

  loss_fn = nn.CrossEntropyLoss()

  for epoch in range(num_epochs):

    # Entrenar el modelo en batches
    for X_batch, y_batch in train_loader:
      y_pred = model(X_batch)
      loss = loss_fn(y_pred, y_batch)
      optimizer.zero_grad()
      loss.backward()
      optimizer.step()

    # Validación
    for X_i, y_i in val_loader:
      y_pred = model(X_i)
      val_loss = loss_fn(y_pred, y_i).item()
      val_accuracy = calc_accuracy(y_pred, y_i)

    print("Epoch: ", epoch+1, f"\tLoss: {val_loss:.4f} \tAccuracy: {val_accuracy:.4f}")


# Implementación del loop de entrenamiento
## Red Neuronal 1

- Capas ocultas: 120
- Épocas: 12
- Batches: 35
- Optimizer: SGD
- Learning Rate: 0.01

In [80]:
# Entrenar el modelo con la clase establecida
model_m1 = FeedForwardNN(784, 120, 10)
optimizer_m1 = optim.SGD(model_m1.parameters(), lr = 0.01)
epochs_m1 = 12
batches_m1 = 35

model_m1.train()
train(model_m1, epochs_m1, optimizer_m1, batches_m1)

Epoch:  1 	Loss: 0.4395 	Accuracy: 0.8828
Epoch:  2 	Loss: 0.3348 	Accuracy: 0.9073
Epoch:  3 	Loss: 0.3107 	Accuracy: 0.9140
Epoch:  4 	Loss: 0.2737 	Accuracy: 0.9226
Epoch:  5 	Loss: 0.2551 	Accuracy: 0.9287
Epoch:  6 	Loss: 0.2463 	Accuracy: 0.9289
Epoch:  7 	Loss: 0.2220 	Accuracy: 0.9380
Epoch:  8 	Loss: 0.2093 	Accuracy: 0.9417
Epoch:  9 	Loss: 0.1997 	Accuracy: 0.9451
Epoch:  10 	Loss: 0.1944 	Accuracy: 0.9454
Epoch:  11 	Loss: 0.1738 	Accuracy: 0.9499
Epoch:  12 	Loss: 0.1695 	Accuracy: 0.9535


# Red Neuronal 2

- Capas ocultas: 240
- Épocas: 10
- Batches: 28
- Optimizer: Adam
- Learning Rate: 0.0075

In [81]:
# Mismo modelo, misma clase, diferentes parámetros
model_m2 = FeedForwardNN(784, 240, 10)
optimizer_m2 = optim.Adam(model_m2.parameters(), lr = 0.0075)
epochs_m2 = 10
batches_m2 = 28


model_m2.train()
train(model_m2, epochs_m2, optimizer_m2, batches_m2)

Epoch:  1 	Loss: 0.1023 	Accuracy: 0.9692
Epoch:  2 	Loss: 0.1082 	Accuracy: 0.9739
Epoch:  3 	Loss: 0.1110 	Accuracy: 0.9683
Epoch:  4 	Loss: 0.0723 	Accuracy: 0.9799
Epoch:  5 	Loss: 0.0838 	Accuracy: 0.9791
Epoch:  6 	Loss: 0.0651 	Accuracy: 0.9815
Epoch:  7 	Loss: 0.1272 	Accuracy: 0.9720
Epoch:  8 	Loss: 0.0742 	Accuracy: 0.9810
Epoch:  9 	Loss: 0.0694 	Accuracy: 0.9824
Epoch:  10 	Loss: 0.0669 	Accuracy: 0.9809


In [82]:
# Clase de Red Neuronal #2, con capas ocultas adicionales
class FeedForwardNN_ALT(nn.Module):
  def __init__(self, input_size, hidden_size, hidden_size_dos, hidden_size_tres, output_size):
    super().__init__()
    self.hidden = nn.Linear(input_size, hidden_size)
    self.hidden2 = nn.Linear(hidden_size, hidden_size_dos)
    self.hidden3 = nn.Linear(hidden_size_dos, hidden_size_tres)
    self.output = nn.Linear(hidden_size_tres, output_size)
    self.relu = nn.ReLU()

  def forward(self, x):
    x = x.reshape(-1, 784)
    x = self.hidden(x)
    x = self.relu(x)
    x = self.hidden2(x)
    x = self.relu(x)
    x = self.hidden3(x)
    x = self.relu(x)
    x = self.output(x)
    return x

# Red Neuronal 3

- Capa oculta 1: 100
- Capa oculta 2: 185
- Capa oculta 3: 190
- Épocas: 9
- Batches: 20
- Optimizer: SGD
- Learning Rate: 0.001

In [83]:
# Clase diferente
model_m3 = FeedForwardNN_ALT(784, 100, 185, 190, 10)
optimizer_m3 = optim.SGD(model_m3.parameters(), lr = 0.001)
epochs_m3 = 9
batches_m3 = 20

model_m3.train()
train(model_m3, epochs_m3, optimizer_m3, batches_m3)

Epoch:  1 	Loss: 2.2834 	Accuracy: 0.4281
Epoch:  2 	Loss: 2.2269 	Accuracy: 0.4903
Epoch:  3 	Loss: 1.9198 	Accuracy: 0.5169
Epoch:  4 	Loss: 1.1175 	Accuracy: 0.7112
Epoch:  5 	Loss: 0.7122 	Accuracy: 0.7926
Epoch:  6 	Loss: 0.5854 	Accuracy: 0.8251
Epoch:  7 	Loss: 0.4965 	Accuracy: 0.8621
Epoch:  8 	Loss: 0.4427 	Accuracy: 0.8734
Epoch:  9 	Loss: 0.4075 	Accuracy: 0.8804


# Mejor modelo obtenido

Ganó la Red Neuronal 2, con un Loss de 0.0669 y un Accuracy de 0.9809.
Mientras que la Red Neuronal 1 obtuvo un Loss de 0.1695 y un Accuracy de 0.9535; y la Red Neuronal 3 obtuvo un Loss de 0.4075 y un Accuracy de 0.8804.
A continuación se evaluará este modelo con valores de prueba.

In [84]:
test_data = DataLoader(mnist_test, batch_size=10000, shuffle=False)
loss_fn = nn.CrossEntropyLoss()

model_m2.eval()

# Evaluación con datos de prueba
with torch.no_grad():

  for X_test, y_test in test_data:

    y_pred = model_m2(X_test)
    loss = loss_fn(y_pred, y_test).item()
    acc = calc_accuracy(y_pred, y_test)


print(f"Test Loss: {loss:.4f}\n")
print(f"Accuracy (Global): {acc:.4f}\n")

Test Loss: 0.2261

Accuracy (Global): 0.9666



# Reflexión personal

Se me hace muy interesante el efecto que tiene cambiar los parámetros de una red neuronal (FF). Por ejemplo, la implementación de varias capas ocultas demostró que no era buena idea para el desempeño del modelo. Eso es una cosa que no podía imaginar. Claro que no todo puede (o tiene que) ser straight-forward; si únicamente con añadir capas ocultas se pudiese mejorar el desempeño de un modelo sin repercusiones, no habría que tener tanta pre-ponderación sobre el tema a tratar. Los datos se han de manejar con cuidado, lógica y raciocinio. Este trabajo me ha enseñado que a veces menos es más, y que tratar de implementar "más" no siempre significa mejor.

Definitivamente pude apreciar (tanto observar como verdaderamente apreciar) cómo es que PyTorch facilita la vida de los analistas de datos (o ingenieros como nosotros).